bindings/go/dag/sync/process.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	if d.LengthVertices() == 0 {
+		return nil
+	}
+	topology := d.Clone()
+	topology, err := topology.Reverse()
+	if err != nil {
+		return fmt.Errorf("failed to reverse graph: %w", err)
+	}
+
+	// Collect all Children nodes that are end leafs
+	roots := topology.Roots()
+	for _, r := range roots {
+		d.MustGetVertex(r).Attributes.Store(AttributeProcessingState, ProcessingStateQueued)
+	}
+
+	// A map to track doneMap nodes
+	doneMap := &sync.Map{}
+
+	// Process nodes concurrently
+	if err := topology.processTopology(ctx, d, roots, processor, doneMap, options); err != nil {
+		return err
+	}
+
+	if topology.LengthVertices() > 0 {
+		return fmt.Errorf("failed to process all objects, remaining: %v", topology.Vertices)
+	}
+
+	return nil
+}
+
+type VertexProcessor[T cmp.Ordered] interface {
+	ProcessVertex(ctx context.Context, vertex T) error
+}
+
+type VertexProcessorFunc[T cmp.Ordered] func(ctx context.Context, vertex T) error
+
+func (f VertexProcessorFunc[T]) ProcessVertex(ctx context.Context, vertex T) error {
+	return f(ctx, vertex)
+}
+
+func (d *DirectedAcyclicGraph[T]) processTopology(
+	ctx context.Context,
+	dag *DirectedAcyclicGraph[T],
+	ids []T, // a list of root nodes to start processing with
+	processor VertexProcessor[T], // the processing function
+	doneMap *sync.Map, // a map to track loaded nodes
+	opts *ProcessTopologyOptions,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	errGroup.SetLimit(opts.GoRoutineLimit)
+	// Calculate the upper bound for the next queue channel
+	// this determines how many ids can be processed concurrently in the next
+	// phase
+	upperBound := 0
+	for _, id := range ids {
+		upperBound += d.MustGetOutDegree(id)
+	}
+
+	nextQueueCh := make(chan T, upperBound)
+
+	for _, id := range ids {
+		errGroup.Go(func() error {
+			// Mark the id as processed or return if already processed.
+			_, loaded := doneMap.LoadOrStore(id, true)
+			if loaded {
+				return nil
+			}
+
+			dag.MustGetVertex(id).Attributes.Store(AttributeProcessingState, ProcessingStateProcessing)
+			if err := processor.ProcessVertex(ctx, id); err != nil {
+				dag.MustGetVertex(id).Attributes.Store(AttributeProcessingState, ProcessingStateError)
+				return fmt.Errorf("failed to process vertex with id %v: %w", id, err)
+			}
+			dag.MustGetVertex(id).Attributes.Store(AttributeProcessingState, ProcessingStateCompleted)
+
+			vertex := d.MustGetVertex(id)
+			for _, parent := range vertex.EdgeKeys() {
+				inDegree := d.MustGetInDegree(parent)
+				d.InDegree.Store(parent, inDegree-1)
+				parentVertex := d.MustGetVertex(parent)
+				parentVertex.Edges.Delete(id)
+				// If all prerequisites (children) of the parent have been
+				// processed and the parent has not been enqueued yet, add it to
+				// the next level.
+				if d.MustGetInDegree(parent) == 0 {
+					dag.MustGetVertex(parent).Attributes.Store(AttributeProcessingState, ProcessingStateQueued)
+					nextQueueCh <- parent
+				}
```

Comment:
This is super hard to follow. This is in degree of `d` but then MustGetVertex is called on `dag`. Should this be `d.MustGetVertex`?